src/RooMorphingPdf.cc

```diff
@@ -0,0 +1,589 @@
+  // PDF, could store as class data members instead
+  // Also, axis1 == axis2 and nb1 == nb2
+  Int_t nb1 = morph_axis_.GetNbins();
+  Int_t nb2 = morph_axis_.GetNbins();
+  Int_t nbn = morph_axis_.GetNbins();
+
+  TArrayD const* bedgesn = morph_axis_.GetXbins();
+
+  Double_t xminn = (*bedgesn)[0];
+  Double_t xmaxn = (*bedgesn)[nbn];
+
+  // ......The weights (wt1,wt2) are the complements of the "distances" between
+  //       the values of the parameters at the histograms and the desired
+  //       interpolation point. For example, wt1=0, wt2=1 means that the
+  //       interpolated histogram should be identical to input histogram 2.
+  //       Check that they make sense. If par1=par2 then we can choose any
+  //       valid set of wt1,wt2 so why not take the average?
+
+  Double_t wt1, wt2;
+  if (par2 != par1) {
+    wt1 = 1. - (parinterp - par1) / (par2 - par1);
+    wt2 = 1. + (parinterp - par2) / (par2 - par1);
+  } else {
+    wt1 = 0.5;
+    wt2 = 0.5;
+  }
+
+  // ......Give a warning if this is an extrapolation.
+
+  if (wt1 < 0 || wt1 > 1. || wt2 < 0. || wt2 > 1. ||
+      fabs(1 - (wt1 + wt2)) > 1.0e-4) {
+    std::cout << "Warning! th1fmorph: This is an extrapolation!! Weights are "
+              << wt1 << " and " << wt2 << " (sum=" << wt1 + wt2 << ")"
+              << std::endl;
+  }
+  if (idebug >= 1)
+    std::cout << "th1morph - Weights: " << wt1 << " " << wt2 << std::endl;
+
+  if (idebug >= 1)
+    std::cout << "New hist: " << nbn << " " << xminn << " " << xmaxn
+              << std::endl;
+
+  // Treatment for empty histograms: Return an empty histogram
+  // with interpolated bins.
+
+  if (hist1.Integral() <= 0 || hist2.Integral() <= 0) {
+    std::cout << "Warning! th1morph detects an empty input histogram. Empty "
+            "interpolated histogram returned: " << std::endl;
+    return (TH1F("morphed", "morphed", nbn, xminn, xmaxn));
+  }
+  if (idebug >= 1)
+    std::cout << "Input histogram content sums: " << hist1.Integral() << " "
+         << hist2.Integral() << std::endl;
+  // *
+  // *......Extract the single precision histograms into double precision arrays
+  // *      for the interpolation computation. The offset is because sigdis(i)
+  // *      describes edge i (there are nbins+1 of them) while dist1/2
+  // *      describe bin i. Be careful, ROOT does not use C++ convention to
+  // *      number bins: dist1[ibin] is content of bin ibin where ibin runs from
+  // *      1 to nbins. We allocate some extra space for the derived
+  // distributions
+  // *      because there may be as many as nb1+nb2+2 edges in the intermediate
+  // *      interpolated cdf described by xdisn[i] (position of edge i) and
+  // *      sigdisn[i] (cummulative probability up this edge) before we project
+  // *      into the final binning.
+  // Float_t const* dist1 = hist1.GetArray();
+  // Float_t const* dist2 = hist2.GetArray();
+
+  // AG: can re-use these too if nb1 and nb2 are fixed
```

Comment:
I'd suggest to re-use the arrays rather than call new and delete within a  function that may be called multiple times within a minimization loop.
Even if `nb1` and `nb2` may not be fixed, but for sure they're bounded from above by a constant known at initialization time (the max number of bins in any template, I think)